[Clang] Eagerly instantiate used constexpr function upon definition.

cor3ntin opened the pull request:
Despite CWG2497 not being resolved, it is reasonable to expect the following code to compile (and which is supported by other compilers)

```cpp
  template<typename T> constexpr T f();
  constexpr int g() { return f<int>(); } // #1
  template<typename T> constexpr T f() { return 123; }
  int k[g()];
  // #2
```

To that end, we eagerly instantiate all referenced specializations of constexpr functions when they are defined.

We maintain a map of (pattern, [instantiations]) independant of `PendingInstantiations` to avoid having to iterate that list after each function definition.

We should apply the same logic to constexpr variables, but I wanted to keep the PR small.

Fixes #73232

cor3ntin committed:
[Clang] Eagerly instantiate used constexpr function upon definition.

Despite CWG2497 not being resolved, it is reasonable to expect the following
code to compile (and which is supported by other compilers)

```cpp
  template<typename T> constexpr T f();
  constexpr int g() { return f<int>(); } // #1
  template<typename T> constexpr T f() { return 123; }
  int k[g()];
  // #2
```

To that end, we eagerly instantiate all referenced
specializations of constexpr functions when they are defined.

We maintain a map of (pattern, [instantiations]) independant of
`PendingInstantiations` to avoid having to iterate that list after
each function definition.

We should apply the same logic to constexpr variables,
but I wanted to keep the PR small.

Fixes #73232

tbaederr commented:
Does this have any compile-time impact?

cor3ntin commented:
> Does this have any compile-time impact?

Nope, should be ~free. It does add 1 DenseMap lookup per constexpr function definition, which ought to be negligible.

zygoloid commented:
I *think* the new map should be unnecessary: `FunctionTemplateDecl` and `VarTemplateDecl` already have a list of specializations that you can walk and check the point of instantiation and template specialization kind to see if there's a pending instantiation.

If you want/need to keep the map approach, you'll need to also deal with the case where the template is declared in a PCH and then defined somewhere that includes the PCH, which will mean serializing and deserializing the map contents.

zygoloid commented:
> > I _think_ the new map should be unnecessary: `FunctionTemplateDecl` and `VarTemplateDecl` already have a list of specializations that you can walk and check the point of instantiation and template specialization kind to see if there's a pending instantiation.
> 
> I don't think that's viable for non-template but templated entities like member function in a class template. Am I missing something?

Hmm, good point, we don't have a way to walk the specializations of a templated function that's not a function template. (You'd need to walk all the instantiations of the enclosing template and check to see if the corresponding member function has a pending instantiation, or something like that, which sounds painful. Friends defined in class templates would probably be especially problematic.)

Let's use your map approach then :-)

cor3ntin committed:
WIP: Serialize pending constexr instantiations

I wanted to avoid loading eagerly all function templates
and their instantiations -
even if ultimately `PendingInstantiations` - which contains the same information
is going to be loaded at the end of the TU.

So instead, when we find the definition of a constexpr
function template we load the associated instantiation from the
external modules at that point.

I'm not sure there is a better way to get to a DeclID from a Decl*
than to iterate over all redeclarations?

I still need to cleanup but I'd like feedback on direction first!

cor3ntin commented:
@zygoloid It took me a while, but I think I got it; what do you think of direction ? (description in the last commit message)
It appears I am doing something a bit novel, but I couldn't think of a better way.

zygoloid commented on clang/include/clang/Sema/ExternalSemaSource.h:
```diff
@@ -181,6 +181,9 @@ class ExternalSemaSource : public ExternalASTSource {
                  SmallVectorImpl<std::pair<ValueDecl *,
                                            SourceLocation> > &Pending) {}
 
+  virtual void ReadPendingOfInstantiationsForConstexprEntity(
```
The "Of" here seems out of place; maybe remove it? You also have an inconsistent "Of" vs "For" after "Instantiations". Would be good to use a consistent spelling.

zygoloid commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -6481,6 +6481,33 @@ void Sema::PerformPendingInstantiations(bool LocalOnly) {
+// definition. This make sure that function template that are defined after the
+// point of instantiation of their used can be evaluated after they are defined.
+// see CWG2497.
+void Sema::InstantiateFunctionTemplateSpecializations(
```
This name doesn't describe what the function does very well. Maybe something more like `PerformPendingInstantiationsOfConstexprFunction` would be clearer?

zygoloid commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -6481,6 +6481,33 @@ void Sema::PerformPendingInstantiations(bool LocalOnly) {
+  auto It =
+      PendingInstantiationsOfConstexprEntities.find(Tpl->getCanonicalDecl());
+  if (It != PendingInstantiationsOfConstexprEntities.end()) {
+    InstantiateAll(It->second);
```
Is this iterator guaranteed to remain valid throughout this call? We might add new keys to the map. Maybe we should move the vector out of the map first.

zygoloid commented on clang/lib/Serialization/ASTReader.cpp:
```diff
@@ -8718,6 +8731,18 @@ void ASTReader::ReadPendingInstantiations(
+    const NamedDecl *D, llvm::SmallSetVector<NamedDecl *, 4> &Decls) {
+  for (auto *Redecl : D->redecls()) {
+    DeclID Id = Redecl->getGlobalID();
+    auto It = PendingInstantiationsOfConstexprEntities.find(Id);
```
Maybe skip this lookup for declarations that aren't from an AST file?

zygoloid commented on clang/lib/Serialization/ASTWriter.cpp:
```diff
@@ -4836,6 +4837,16 @@ ASTFileSignature ASTWriter::WriteASTCore(Sema &SemaRef, StringRef isysroot,
   assert(SemaRef.PendingLocalImplicitInstantiations.empty() &&
          "There are local ones at end of translation unit!");
 
+  // Build a record containing all of pending instantiations of constexpr
```
```suggestion
  // Build a record containing all pending instantiations of constexpr
```

zygoloid commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -6481,6 +6481,33 @@ void Sema::PerformPendingInstantiations(bool LocalOnly) {
+
+  auto InstantiateAll = [&](const auto &Range) {
+    for (NamedDecl *Fun : Range) {
+      InstantiateFunctionDefinition(PointOfInstantiation,
```
I think the point of instantiation should be read from the FunctionDecl, not the point where the function is defined, so that the "in instantiation of" note points to the right place.

zygoloid reviewed (commented):
Thanks!

cor3ntin committed:
Address Richard's feedback.

zygoloid commented on clang/test/SemaCXX/cxx2b-consteval-propagate.cpp:
```diff
@@ -112,6 +112,9 @@ template <typename T>
 constexpr int f(T t) { // expected-note {{'f<int>' defined here}}
     return id(t); // expected-note {{'f<int>' is an immediate function because its body contains a call to a consteval function 'id' and that call is not a constant expression}}
 }
+
+
+
```
Revert these added blank lines.

zygoloid reviewed (approved):
Thanks! I assume you'll also be looking at the variable case (perhaps as a separate commit)?

shafik reviewed (approved):
Thank you, LGTM

cor3ntin committed:
Remove new lines

cor3ntin commented:
> Thanks! I assume you'll also be looking at the variable case (perhaps as a separate commit)?

I'll try to get to it, yes!
Thanks for the review

sam-mccall commented on clang/lib/Serialization/ASTWriter.cpp:
```diff
@@ -4836,6 +4837,16 @@ ASTFileSignature ASTWriter::WriteASTCore(Sema &SemaRef, StringRef isysroot,
+  // Build a record containing all pending instantiations of constexpr
+  // entities.
+  RecordData PendingInstantiationsOfConstexprEntities;
+  for (const auto &I : SemaRef.PendingInstantiationsOfConstexprEntities) {
```
In case this will be relanded in some form:

This loop uses arbitrary hashtable order, leading to nondeterministic PCM output.
